### DISCUSSION TITLE
Which Branch Should I Download?

### DISCUSSION THREAD

**Discussion by Curry-Icecream** [brgl/libgpiod#58] [Q&A]
Currently, Im using raspberry pi 5, should I download v2.1 or the master branch?

Currently what I'm doing is I've downloaded the master branch and did what have been told in ReadMe file, then put the bin, include and lib into /usr/local/"corresponding dir", then I used Make All in geany and tried to compile the example code (the one about toggle line), and it did say pin 17 is active and inactive consecutively, but my led connected to pin 17 didn't do anything.

If I build instead of Make, then it will say undefined to reference to something.

**Comment by brgl** [comment]
> Currently, Im using raspberry pi 5, should I download v2.1 or the master branch?

You should typically use the latest stable release which is now v2.1.

> Currently what I'm doing is I've downloaded the master branch and did what have been told in ReadMe file, then put the bin, include and lib into /usr/local/"corresponding dir", then I used Make All in geany and tried to compile the example code (the one about toggle line), and it did say pin 17 is active and inactive consecutively, but my led connected to pin 17 didn't do anything.
> 

This is way too little information to help. When submitting a bug report you need to give developers as much info as possible. Exact commands you run, program output, environment etc.

> If I build instead of Make, then it will say undefined to reference to something.

Same as above. I have no idea what `undefined reference to something` means.

**Comment by Curry-Icecream** [comment]
Thanks for the reply.
I have re-downloaded the library v2.1.x and tried to build "toggle_line_value.c" on geany and it gives me the following message:

gcc -Wall -o "toggle_line_value" "toggle_line_value.c" -lgpiod (in directory: /home/uobgip1/project/libgpiod/examples)
/usr/bin/ld: /tmp/cc3Gekjp.o: in function `request_output_line':
toggle_line_value.c:(.text+0x40): undefined reference to `gpiod_line_settings_new'
/usr/bin/ld: toggle_line_value.c:(.text+0x5c): undefined reference to `gpiod_line_settings_set_direction'
/usr/bin/ld: toggle_line_value.c:(.text+0x68): undefined reference to `gpiod_line_settings_set_output_value'
/usr/bin/ld: toggle_line_value.c:(.text+0x6c): undefined reference to `gpiod_line_config_new'
/usr/bin/ld: toggle_line_value.c:(.text+0x94): undefined reference to `gpiod_line_config_add_line_settings'
/usr/bin/ld: toggle_line_value.c:(.text+0xb4): undefined reference to `gpiod_request_config_new'
/usr/bin/ld: toggle_line_value.c:(.text+0xd0): undefined reference to `gpiod_request_config_set_consumer'
/usr/bin/ld: toggle_line_value.c:(.text+0xe0): undefined reference to `gpiod_chip_request_lines'
/usr/bin/ld: toggle_line_value.c:(.text+0xec): undefined reference to `gpiod_request_config_free'
/usr/bin/ld: toggle_line_value.c:(.text+0x104): undefined reference to `gpiod_line_config_free'
/usr/bin/ld: toggle_line_value.c:(.text+0x114): undefined reference to `gpiod_line_settings_free'
/usr/bin/ld: /tmp/cc3Gekjp.o: in function `main':
toggle_line_value.c:(.text+0x274): undefined reference to `gpiod_line_request_set_value'
collect2: error: ld returned 1 exit status
Compilation failed.

**Reply by brgl** [reply]
The project - as per README - is built using autotools. Please use them as I don't support any other way of doing it.

**Comment by Curry-Icecream** [comment]
So if I choose Make all, then I can compile the code and the terminal would write:
17=Active
17=Inactive
17=Active

I have connected a led light to gpio 17, but it doesn't turn on when 17=Active.
The only thing I have changed is static const unsigned int line_offset = 17;, the other things are same as the example code "toggle_line_value.c"

**Reply by brgl** [reply]
This is still not enough information. Run the program through `strace` and post the output here.
